src/Symfony/Component/DependencyInjection/EnvProvider.php

```diff
@@ -0,0 +1,150 @@
+            return preg_replace_callback('/%%|%([^%\s]+)%/', function ($match) use ($name) {
+                if (!isset($match[1])) {
+                    return '%';
+                }
+                if (!is_scalar($value = $this->container->getParameter($match[1]))) {
+                    throw new RuntimeException(sprintf('Parameter "%s" found when resolving env var "%s" must be scalar, "%s" given.', $match[1], $name, gettype($value)));
+                }
+
+                return $value;
+            }, self::scalar($getEnv($name), $name, $prefix));
+        }
+
+        $i = strpos($name, ':');
+
+        if ('file' === $prefix) {
+            if (0 < $i && strspn($name, '0123456789') === $i) {
+                $maxLength = (int) $name;
+                $name = substr($name, 1 + $i);
+            } else {
+                $maxLength = null;
+            }
+            if (!file_exists($file = self::scalar($getEnv($name), $name, $prefix))) {
+                throw new RuntimeException(sprintf('Env "file:%s" not found: %s does not exist.', $name, $file));
+            }
+
+            return null === $maxLength ? file_get_contents($file) : file_get_contents($file, false, null, 0, $maxLength);
+        }
+
+        if (false !== $i) {
+            $env = $getEnv($name);
+        } elseif (0 !== strpos($name, 'HTTP_') && isset($_SERVER[$name])) {
+            $env = $_SERVER[$name];
+        } elseif (isset($_ENV[$name])) {
+            $env = $_ENV[$name];
+        } elseif (false === $env = getenv($name)) {
+            return;
+        }
+        $env = self::scalar($env, $name, $prefix);
+
+        if ('string' === $prefix) {
+            return (string) $env;
+        }
+
+        if ('const' === $prefix) {
+            if (!defined($env)) {
+                throw new RuntimeException(sprintf('Env var "%s" maps to undefined constant "%s".', $name, $env));
+            }
+
+            return constant($name);
+        }
+
+        if ('bool' === $prefix) {
+            return (bool) XmlUtils::phpize($env);
+        }
+
+        if ('int' === $prefix) {
+            if (!is_numeric($env = XmlUtils::phpize($env))) {
+                throw new RuntimeException(sprintf('Non-numeric env var "%s" cannot be cast to int.', $name));
+            }
+
+            return (int) $env;
+        }
+
+        if ('float' === $prefix) {
+            if (!is_numeric($env = XmlUtils::phpize($env))) {
+                throw new RuntimeException(sprintf('Non-numeric env var "%s" cannot be cast to float.', $name));
+            }
+
+            return (float) $env;
+        }
+
+        throw new RuntimeException(sprintf('Unsupported env var prefix "%s".', $prefix));
+    }
+
+    final public static function scalar($env, $name, $prefix)
+    {
+        if (!is_scalar($env)) {
+            throw new RuntimeException(sprintf('Non-scalar env var "%s" cannot be cast to %s.', $name, $prefix));
+        }
+
+        return $env;
+    }
+}
```

Comment:
And then asking the interface to support only one prefix actually rather than an array of providers... ?